### ISSUE TITLE
[Shell] Items - Binding Visible

### ISSUE THREAD

**Issue by dazinator** [xamarin/Xamarin.Forms#5428]
## Summary
I have installed VS2019 and had a go with the new Xamarin Forms Shell templates.

One area which I think is lacking is the ability to bind `ShellItem` `ShellSection` visibility.

For example, one common use case, is, based on the authenticated user's permissions, certain items on the menu should be visible, or not visible. Permissions of the user could be changed during application use.

The shell spec allows you to define your nav menu by defining `ShellItem`s and `ShellSection`s.
However I cannot see anyway to bind their visibility to fulfill the above use case.

## API Changes
I am new to this spec so its possible I have overlooked something. However I believe that the Shell components should have a Visibility property that can be bound through ordinary data binding mechanism.

## Intended Use Case
As per above, main use case is to dynamically allow menu items to appear or disappear based on for example, security context of current user.
Parent: #2415

**Comment by dbwelch** [comment]
Thanks for opening the issue!  Perfect! @dazinator

**Comment by dbwelch** [comment]
Just to add to this, the items themselves should be able to be added/removed, as some content (in my applications case) is shown to based upon their user class.  

So, the parameters passed to the content (in the template) will be different.

**Comment by fhnainia** [comment]
I am also having issues dynamically adding ShellContent items to an existing ShellSection. In my case it seems to work well in Android but does not work in iOS!
Here is how I am attempting to do this.
1. In the Shell Page I declare a ShellSection as follows:

        <ShellSection x:Name="Halls" Title="Halls and Tables" Icon="tab_feed.png">
                 <ShellContent ContentTemplate="{DataTemplate local:Halls}" ></ShellContent>
        </ShellSection>

2. In the Halls.xaml.cs page I attempt to load the Halls and Tables dynamically. Each Hall is supposed to be ShellContent item.

               var shell = App.Current.MainPage as Shell;

                ShellSection section = (ShellSection)shell.CurrentItem.CurrentItem;
                section.Items.Clear();
                section.Title = "Halls & Tables";

                if (viewModel.Items.Count == 0)
                {
                    viewModel.LoadItemsCommand.Execute(null);
                }

                foreach (var item in viewModel.Items)
                {
                    ShellContent sContent = new ShellContent()
                    {
                        Title = item.Name,
                        Content = new Tables(item)
                        
                    };
                    section.Items.Add(sContent);
                }

This works well in Android but it crashes in IOS with the following:

System.NullReferenceException: Object reference not set to an instance of an object

2019-03-02 16:42:23.659 SkiaSharpShellDemo.iOS[11377:3934923] 
Unhandled Exception:
System.NullReferenceException: Object reference not set to an instance of an object
  at Xamarin.Forms.Platform.iOS.ShellSectionRootHeader.LayoutBar () [0x00030] in D:\a\1\s\Xamarin.Forms.Platform.iOS\Renderers\ShellSectionRootHeader.cs:197 
  at Xamarin.Forms.Platform.iOS.ShellSectionRootHeader.UpdateSelectedIndex (System.Boolean animated) [0x00022] in D:\a\1\s\Xamarin.Forms.Platform.iOS\Renderers\ShellSectionRootHeader.cs:220 
  at Xamarin.Forms.Platform.iOS.ShellSectionRootHeader.ViewDidLoad () [0x001a2] in D:\a\1\s\Xamarin.Forms.Platform.iOS\Renderers\ShellSectionRootHeader.cs:172 
  at (wrapper managed-to-native) ObjCRuntime.Messaging.IntPtr_objc_msgSendSuper(intptr,intptr)
  at UIKit.UIViewController.get_View () [0x0002a] in /Library/Frameworks/Xamarin.iOS.framework/Versions/12.6.0.23/src/Xamarin.iOS/UIKit/UIViewController.g.cs:3037 
  at Xamarin.Forms.Platform.iOS.ShellSectionRootRenderer.UpdateHeaderVisibility () [0x00
04b] in D:\a\1\s\Xamarin.Forms.Platform.iOS\Renderers\ShellSectionRootRenderer.cs:206 
  at Xamarin.Forms.Platform.iOS.ShellSectionRootRenderer.ViewDidLoad () [0x0009f] in D:\a\1\s\Xamarin.Forms.Platform.iOS\Renderers\ShellSectionRootRenderer.cs:70 
  at (wrapper managed-to-native) UIKit.UIApplication.UIApplicationMain(int,string[],intptr,intptr)
  at UIKit.UIApplication.Main (System.String[] args, System.IntPtr principal, System.IntPtr delegate) [0x00005] in /Library/Frameworks/Xamarin.iOS.framework/Versions/12.6.0.23/src/Xamarin.iOS/UIKit/UIApplication.cs:79 
  at UIKit.UIApplication.Main (System.String[] args, System.String principalClassName, System.String delegateClassName) [0x0002c] in /Library/Frameworks/Xamarin.iOS.framework/Versions/12.6.0.23/src/Xamarin.iOS/UIKit/UIApplication.cs:63 
  at SkiaSharpShellDemo.iOS.Application.Main (System.String[] args) [0x00001] in C:\Users\fhnainia\source\repos\SkiaSharpShellDemo\SkiaSharpShellDemo\SkiaSharpShellDemo.iOS\Main.cs:17 
2019-03-02 16:42:23.661 SkiaSharpShellDemo.iOS[11377:3934923] Unhandled managed exception:
Object reference not set to an instance of an object (System.NullReferenceException)
  at Xamarin.Forms.Platform.iOS.ShellSectionRootHeader.LayoutBar () [0x00030] in D:\a\1\s\Xamarin.Forms.Platform.iOS\Renderers\ShellSectionRootHeader.cs:197 
  at Xamarin.Forms.Platform.iOS.ShellSectionRootHeader.UpdateSelectedIndex (System.Boolean animated) [0x00022] in D:\a\1\s\Xamarin.Forms.Platform.iOS\Renderers\ShellSectionRootHeader.cs:220 
  at Xamarin.Forms.Platform.iOS.ShellSectionRootHeader.ViewDidLoad () [0x001a2] in D:\a\1\s\Xamarin.Forms.Platform.iOS\Renderers\ShellSectionRootHeader.cs:172 
  at (wrapper managed-to-native) ObjCRuntime.Messaging.IntPtr_objc_msgSendSuper(intptr,intptr)
  at UIKit.UIViewController.get_View () [0x0002a] in /Library/Frameworks/Xamarin.iOS.framework/Versions/12.6.0.23/src/Xamarin.iOS/UIKit/UIViewController.g.cs:3037 
  at Xamarin.Forms.Platform.iOS.ShellSectionRootRenderer.UpdateHeaderVisibility 
The app has been terminated.() [0x0004b] in D:\a\1\s\Xamarin.Forms.Platform.iOS\Renderers\ShellSectionRootRenderer.cs:206 
  at Xamarin.Forms.Platform.iOS.ShellSectionRootRenderer.ViewDidLoad () [0x0009f] in D:\a\1\s\Xamarin.Forms.Platform.iOS\Renderers\ShellSectionRootRenderer.cs:70 
  at (wrapper managed-to-native) UIKit.UIApplication.UIApplicationMain(int,string[],intptr,intptr)
  at UIKit.UIApplication.Main (System.String[] args, System.IntPtr principal, System.IntPtr delegate) [0x00005] in /Library/Frameworks/Xamarin.iOS.framework/Versions/12.6.0.23/src/Xamarin.iOS/UIKit/UIApplication.cs:79 
  at UIKit.UIApplication.Main (System.String[] args, System.String principalClassName, System.String delegateClassName) [0x0002c] in /Library/Frameworks/Xamarin.iOS.framework/Versions/12.6.0.23/src/Xamarin.iOS/UIKit/UIApplication.cs:63 
  at SkiaSharpShellDemo.iOS.Application.Main (System.String[] args) [0x00001] in C:\Users\fh2020\source\repos\SkiaSharpShellDemo\SkiaSharpShellDemo\SkiaSharpShellDemo.iOS\Main.cs:17

**Comment by fhnainia** [comment]
I am also having issues dynamically adding ShellContent items to an existing ShellSection. In my case it seems to work well in Android but does not work in iOS!
Here is how I am attempting to do this.
1. In the Shell Page I declare a ShellSection as follows:

        <ShellSection x:Name="Halls" Title="Halls and Tables" Icon="tab_feed.png">
                 <ShellContent ContentTemplate="{DataTemplate local:Halls}" ></ShellContent>
        </ShellSection>

2. In the Halls.xaml.cs page I attempt to load the Halls and Tables dynamically. Each Hall is supposed to be ShellContent item.

               var shell = App.Current.MainPage as Shell;

                ShellSection section = (ShellSection)shell.CurrentItem.CurrentItem;
                section.Items.Clear();
                section.Title = "Halls & Tables";

                if (viewModel.Items.Count == 0)
                {
                    viewModel.LoadItemsCommand.Execute(null);
                }

                foreach (var item in viewModel.Items)
                {
                    ShellContent sContent = new ShellContent()
                    {
                        Title = item.Name,
                        Content = new Tables(item)
                        
                    };
                    section.Items.Add(sContent);
                }

This works well in Android but it crashes in IOS with the following:

System.NullReferenceException: Object reference not set to an instance of an object

2019-03-02 16:42:23.659 SkiaSharpShellDemo.iOS[11377:3934923] 
Unhandled Exception:
System.NullReferenceException: Object reference not set to an instance of an object
  at Xamarin.Forms.Platform.iOS.ShellSectionRootHeader.LayoutBar () [0x00030] in D:\a\1\s\Xamarin.Forms.Platform.iOS\Renderers\ShellSectionRootHeader.cs:197 
  at Xamarin.Forms.Platform.iOS.ShellSectionRootHeader.UpdateSelectedIndex (System.Boolean animated) [0x00022] in D:\a\1\s\Xamarin.Forms.Platform.iOS\Renderers\ShellSectionRootHeader.cs:220 
  at Xamarin.Forms.Platform.iOS.ShellSectionRootHeader.ViewDidLoad () [0x001a2] in D:\a\1\s\Xamarin.Forms.Platform.iOS\Renderers\ShellSectionRootHeader.cs:172 
  at (wrapper managed-to-native) ObjCRuntime.Messaging.IntPtr_objc_msgSendSuper(intptr,intptr)
  at UIKit.UIViewController.get_View () [0x0002a] in /Library/Frameworks/Xamarin.iOS.framework/Versions/12.6.0.23/src/Xamarin.iOS/UIKit/UIViewController.g.cs:3037 
  at Xamarin.Forms.Platform.iOS.ShellSectionRootRenderer.UpdateHeaderVisibility () [0x00
04b] in D:\a\1\s\Xamarin.Forms.Platform.iOS\Renderers\ShellSectionRootRenderer.cs:206 
  at Xamarin.Forms.Platform.iOS.ShellSectionRootRenderer.ViewDidLoad () [0x0009f] in D:\a\1\s\Xamarin.Forms.Platform.iOS\Renderers\ShellSectionRootRenderer.cs:70 
  at (wrapper managed-to-native) UIKit.UIApplication.UIApplicationMain(int,string[],intptr,intptr)
  at UIKit.UIApplication.Main (System.String[] args, System.IntPtr principal, System.IntPtr delegate) [0x00005] in /Library/Frameworks/Xamarin.iOS.framework/Versions/12.6.0.23/src/Xamarin.iOS/UIKit/UIApplication.cs:79 
  at UIKit.UIApplication.Main (System.String[] args, System.String principalClassName, System.String delegateClassName) [0x0002c] in /Library/Frameworks/Xamarin.iOS.framework/Versions/12.6.0.23/src/Xamarin.iOS/UIKit/UIApplication.cs:63 
  at SkiaSharpShellDemo.iOS.Application.Main (System.String[] args) [0x00001] in C:\Users\fhnainia\source\repos\SkiaSharpShellDemo\SkiaSharpShellDemo\SkiaSharpShellDemo.iOS\Main.cs:17 
2019-03-02 16:42:23.661 SkiaSharpShellDemo.iOS[11377:3934923] Unhandled managed exception:
Object reference not set to an instance of an object (System.NullReferenceException)
  at Xamarin.Forms.Platform.iOS.ShellSectionRootHeader.LayoutBar () [0x00030] in D:\a\1\s\Xamarin.Forms.Platform.iOS\Renderers\ShellSectionRootHeader.cs:197 
  at Xamarin.Forms.Platform.iOS.ShellSectionRootHeader.UpdateSelectedIndex (System.Boolean animated) [0x00022] in D:\a\1\s\Xamarin.Forms.Platform.iOS\Renderers\ShellSectionRootHeader.cs:220 
  at Xamarin.Forms.Platform.iOS.ShellSectionRootHeader.ViewDidLoad () [0x001a2] in D:\a\1\s\Xamarin.Forms.Platform.iOS\Renderers\ShellSectionRootHeader.cs:172 
  at (wrapper managed-to-native) ObjCRuntime.Messaging.IntPtr_objc_msgSendSuper(intptr,intptr)
  at UIKit.UIViewController.get_View () [0x0002a] in /Library/Frameworks/Xamarin.iOS.framework/Versions/12.6.0.23/src/Xamarin.iOS/UIKit/UIViewController.g.cs:3037 
  at Xamarin.Forms.Platform.iOS.ShellSectionRootRenderer.UpdateHeaderVisibility 
The app has been terminated.() [0x0004b] in D:\a\1\s\Xamarin.Forms.Platform.iOS\Renderers\ShellSectionRootRenderer.cs:206 
  at Xamarin.Forms.Platform.iOS.ShellSectionRootRenderer.ViewDidLoad () [0x0009f] in D:\a\1\s\Xamarin.Forms.Platform.iOS\Renderers\ShellSectionRootRenderer.cs:70 
  at (wrapper managed-to-native) UIKit.UIApplication.UIApplicationMain(int,string[],intptr,intptr)
  at UIKit.UIApplication.Main (System.String[] args, System.IntPtr principal, System.IntPtr delegate) [0x00005] in /Library/Frameworks/Xamarin.iOS.framework/Versions/12.6.0.23/src/Xamarin.iOS/UIKit/UIApplication.cs:79 
  at UIKit.UIApplication.Main (System.String[] args, System.String principalClassName, System.String delegateClassName) [0x0002c] in /Library/Frameworks/Xamarin.iOS.framework/Versions/12.6.0.23/src/Xamarin.iOS/UIKit/UIApplication.cs:63 
  at SkiaSharpShellDemo.iOS.Application.Main (System.String[] args) [0x00001] in C:\Users\fh2020\source\repos\SkiaSharpShellDemo\SkiaSharpShellDemo\SkiaSharpShellDemo.iOS\Main.cs:17

**Comment by dbwelch** [comment]
Any update to this?  Really don't understand how this feature can be used at all without allowing the developer to add/remove/hide/show ShellItems.  Do we have to deploy a new app every time there is a new category/ShellSection?  Do we need a separate app for authenticated users to be able to show them authorized items?  Silly, right?

**Comment by jaluka** [comment]
My Use Case

User is not logged in so need items not to be shown or show disabled.
User logs in items are now shown or enabled.

**Comment by PureWeen** [comment]
@jaluka if you define your layout like this

```XAML
<TabBar>
-LoginPage
</TabBar>

<FlyoutItem Route="LoggedIn">
-LoggedInPage
</FlouyItem>

```


That should give you what you want

- TabBar disables the flyout so it's not visible
- after logging in do `GoToAsync("//LoggedIn")

**Comment by samhouts** [comment]
Duplicate of #5232

**Comment by TheBaileyBrew** [comment]
From what I've discovered @fhnainia - the bulk of the Shell tweaking from the code behind has to be done on AppShell.xaml.cs

I managed to build my app to build all the FlyoutItems dynamically in code, and then add/remove them based on some external consideration.

It goes something like this:
1. Build a List<> of all possible FlyoutItems using a custom Model for uniformity
2. Add items from your List<> using System.Linq and Xamarin.Forms.Internals
3. Enumerate through the created list and remove ShellItems that shouldn't be there with Items.Remove().

Here's a [Gist](https://gist.github.com/TheBaileyBrew/f8a9d2e4668da3ec9bff9bf86d32d951) of what I managed to get working in my app until ShellItem.IsVisible is up and running.
